azerothcore/azerothcore-wotlk
# fix(Core/SmartAI) : use explicit stack DS for ProcessAction instead of recursion

## Changes Proposed:
-  Add a deque to SmartScript object to store data about linked events instead of calling ```SmartScript::ProcessEvent``` recursively
-  ```SmartScript::ProcessEventsFor``` uses the deque to process events

## Issues Addressed:

- Closes https://github.com/azerothcore/azerothcore-wotlk/issues/16674

## SOURCE:

Before:

https://github.com/azerothcore/azerothcore-wotlk/assets/19265585/f7484f17-a606-44f3-a6fa-ad77d5322ace

After:

No longer crashes

## Tests Performed:

- Went to Gundrak and the server did not crash
- Killed Tormar Frostgut in garm and game did not crash
- Went to Shattered Halls to see if mobs execute the scripts correctly and it was ok

## How to Test the Changes:

Run in Debug, MSVC (I have 19.37.32705.0 installed)

Scenario 1:

1. Fly to Gundrak 
2. Should no longer crash
3.

Scenario 2(video):

1. addItem 41431
2. go to end of garm in Storm Peaks
3. place dynamite
4. Kill mob
5. Script should run without crashing

## Known Issues and TODO List:


- [Might affect performance]
- [ I cannot test every script to check if anything is affected]


## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.
